src/aot/aot.cpp

```diff
@@ -146,44 +151,102 @@ int generate(const RequiredResources &resources,
     .bc_len = serialized_bytecode.size(),
   };
 
-  std::ofstream outf(out, std::ios::binary | std::ios::out | std::ios::trunc);
-  if (outf.fail())
-  {
-    LOG(ERROR) << "Failed to open: " << out;
-    return 1;
-  }
+  // Resize the output buffer appropriately
+  std::vector<uint8_t> out;
+  out.resize(sizeof(Header) + hdr.rr_len + hdr.bc_len);
+  uint8_t *p = out.data();
 
   // Write out header
-  outf.write(reinterpret_cast<const char *>(&hdr), sizeof(hdr));
-  if (outf.fail())
+  memcpy(p, &hdr, sizeof(Header));
+  p += sizeof(Header);
+
+  // Write out metadata
+  memcpy(p, serialized_metadata.data(), hdr.rr_len);
+  p += hdr.rr_len;
+
+  // Write out bytecode
+  memcpy(p, serialized_bytecode.data(), hdr.bc_len);
+  p += hdr.bc_len;
+
+  return out;
+}
+
+// Clones the shim to final destination while also injecting
+// the custom .btaot section.
+bool build_binary(const std_filesystem::path &shim,
+                  const std::string &out,
+                  const std::vector<uint8_t> &section)
+{
+  std::error_code ec;
+  bool ret = false;
+  char cmd[1024];
+  int written;
+
+  // Write out section data in a temporary file
+  std::ofstream secdata(AOT_SECDATA_TEMPFILE, std::ios_base::binary);
+  secdata.write(reinterpret_cast<const char *>(section.data()), section.size());
+  secdata.close();
+
+  // Respect user provided BPFTRACE_OBJCOPY if present
+  std::string objcopy = "objcopy";
+  if (auto c = std::getenv("BPFTRACE_OBJCOPY"))
```

Comment:
Should this get added to the list of env vars in both the reference doc and man page:
https://github.com/iovisor/bpftrace/blob/master/docs/reference_guide.md#9-environment-variables

Reply:
My plan was to leave this (and `--aot` cli flag) undocumented until it's ready to be used. Not necessarily completely done, but until at it's at least testable.